flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java

```diff
@@ -300,6 +291,32 @@ private void deleteDirectory(File directory) throws IOException {
 		}
 	}
 
+	/**
+	 * Pruning the useless checkpoints.
+	 */
+	private void pruneCheckpoints(long checkpointID, boolean breakTheIteration) {
+
+		Iterator<Map.Entry<Long, TaskStateSnapshot>> entryIterator =
+			storedTaskStateByCheckpointID.entrySet().iterator();
+
+		final List<Map.Entry<Long, TaskStateSnapshot>> toRemove = new ArrayList<>();
+
+		while (entryIterator.hasNext()) {
+
+			Map.Entry<Long, TaskStateSnapshot> snapshotEntry = entryIterator.next();
+			long entryCheckpointId = snapshotEntry.getKey();
+
+			if (entryCheckpointId != checkpointID) {
```

Comment:
After a second though, while I think this code is currently correct, the case with breaking looks a bit dangerous. Potentially, if the checkpoint id is not there, this would not stop and prune ongoing checkpoints. I wonder if we should make the `if` a bit more complex, but safer (checking that the breaking case never exceeds the checkpoint id). What do you think?

Reply:
I agree with you that the breaking case looks a bit dangerous ... I think maybe we could pass a `Predicate` for the `if` and let the caller side pass the `Predicate` into this function. This could make it cleaner from the caller side and don't need to mass the logic into the `if` to make it complex.

Reply:
That is fine, from my point of view that is just one way of making the `if` more complex.